File: able.js
var exec = require('child_process').exec;
var serial = require('serialport').SerialPort;
var serialport = new serial("/dev/ttyS2", {
    baudrate: 9600
});

var sensor_1 = {
        exec: exec("sudo NOBLE_HCI_DEVICE_ID=0 nodejs scan.js"),
        current: "",
        rssi: []
    },
    sensor_2 = {
        exec: exec("sudo NOBLE_HCI_DEVICE_ID=1 nodejs scan.js"),
        current: "",
        rssi: []
    },
    sensor_3 = {
        exec: exec("sudo NOBLE_HCI_DEVICE_ID=2 nodejs scan.js"),
        current: "",
        rssi: []
    };
var AIL = "s0",
    ELE = "s1",
    THR = "s2",
    RUD = "s3";

function calcDistance(rssi) {
    var focal_point = -16,
        distance;
    distance = Math.pow(10, (focal_point + rssi) / (10 * 2));
    return parseInt(distance);
}

function triangulation(distance_from_beacon_to_a, distance_from_beacon_to_b, distance_from_beacon_to_c, distance_between) {
    function triangle(a, b, c) {
        temp = (Math.acos((Math.pow(a, 2) + Math.pow(c, 2) - Math.pow(b, 2)) / (2 * a * c))) * (180 / Math.PI);

        if (temp > 90) {
            return "left";
        } else if (temp <= 90 && temp >= 45) {
            return "middle";
        } else if (temp < 45) {
            return = "right";
        }
    }

    var front = triangle(distance_from_beacon_to_a, distance_from_beacon_to_b, distance_between);
    var left = triangle(distance_from_beacon_to_a, distance_from_beacon_to_c, distance_between);

    if (front === "left" && left === "right") {
        return "Quadrant 1";
        serialport.open(function (error) {
            if (error) {
                console.log('failed to open: ' + error);
            } else {
                serialport.write(ELE + " 1000", function (err, results) {
                    if (err) {
                        console.log(err);
                    } else if (results) {
                        console.log(results);
                    }
                });
                setTimeout(function () {
                    serialport.write(ELE + " 0", function (err, results) {
                        if (err) {
                            console.log(err);
                        } else if (results) {
                            console.log(results);
                        }
                    });
                }, 5000);

            }
        });
    } else if (front === "left" && left === "middle") {
        return "Quadrant 4";
    } else if (front === "left" && left === "left") {
        return "Quadrant 7";
    } else if (front === "middle" && left === "right") {
        return "Quadrant 2";
    } else if (front === "middle" && left === "middle") {
        return "Quadrant 5";
    } else if (front === "middle" && left === "left") {
        return "Quadrant 8";
    } else if (front === "right" && left === "right") {
        return "Quadrant 3";
    } else if (front === "right" && left === "middle") {
        return "Quadrant 6";
    } else if (front === "right" && left === "left") {
        return "Quadrant 9";
    }
}

console.log("App Start");

sensor_1.exec.stdout.on('data', function (data) {
    if (sensor_1.rssi.length < 10) {
        sensor_1.rssi.push(Math.abs(data));
    } else {
        var minimum = Math.min.apply(Math, sensor_1.rssi);
        sensor_1.rssi.splice(sensor_1.rssi.indexOf(minimum), 1);
        var newLocation = Math.min.apply(Math, sensor_1.rssi);
        sensor_1.current = calcDistance(newLocation);
        console.log("Sensor 1 Location: " + sensor_1.current + " cm");
        sensor_1.rssi = [];
    }
});
sensor_2.exec.stdout.on('data', function (data) {
    if (sensor_2.rssi.length < 10) {
        sensor_2.rssi.push(Math.abs(data));
    } else {
        var minimum = Math.min.apply(Math, sensor_2.rssi);
        sensor_2.rssi.splice(sensor_2.rssi.indexOf(minimum), 1);
        var newLocation = Math.min.apply(Math, sensor_2.rssi);
        sensor_2.current = calcDistance(newLocation);
        console.log("Sensor 2 Location: " + sensor_2.current + " cm");
        sensor_2.rssi = [];
    }
});
sensor_3.exec.stdout.on('data', function (data) {
    if (sensor_3.rssi.length < 10) {
        sensor_3.rssi.push(Math.abs(data));
    } else {
        var minimum = Math.min.apply(Math, sensor_3.rssi);
        sensor_3.rssi.splice(sensor_3.rssi.indexOf(minimum), 1);
        var newLocation = Math.min.apply(Math, sensor_3.rssi);
        sensor_3.current = calcDistance(newLocation);
        console.log("Sensor 3 Location: " + sensor_3.current + " cm");
        sensor_3.rssi = [];
    }
});

setInterval(function () {
    if (sensor_1.current && sensor_2.current && sensor_3.current) {
        triangulation(sensor_1.current, sensor_2.current, sensor_3.current, 30);
        sensor_1.current = "";
        sensor_2.current = "";
        sensor_3.current = "";
    }
}, 1000);

File: able.py
import blescan
import sys
import math
import time
import serial

import bluetooth._bluetooth as bluez

s = 100 #serial.Serial("/dev/ttyS2",9600)

def calcDistance(rssi):
        focal_point = -16
        distance = math.pow(10, (focal_point - rssi) / (10 * 2))
        return float("{0:.2f}".format(distance))

def triangle(a,b,c):
    temp = (math.acos((math.pow(a, 2) + math.pow(c, 2) - math.pow(b, 2)) / (2 * a * c))) * (180 / math.pi)
    if temp > 90:
        return ["left",(math.cos(180 - temp) * a) + 15,math.tan(180 - temp) * a]
    elif temp <= 90 and temp >= 45:
        return ["middle",(math.cos(temp) * a),math.tan(temp) * a]
    elif temp < 45:
        return ["right",(math.cos(temp) * a),math.tan(temp) * a]

def triangulation(distance_from_beacon_to_a, distance_from_beacon_to_b, distance_from_beacon_to_c, distance_between):
    default_position = [2500, 15, 1500]
    front = triangle(distance_from_beacon_to_a, distance_from_beacon_to_b, distance_between)
    left = triangle(distance_from_beacon_to_a, distance_from_beacon_to_c, distance_between)

    avg_z = Math.round((front[2] * left[2]) / 2);

    if avg_z < default_position[0]:
        s.write("st\n") # Servo test command
        time.sleep(5) # Wait for 5 seconds
        s.write("serr\n") # Wrong command to stop servo test
    elif avg_z > default_position[0]:
        s.write("st\n") # Servo test command
        time.sleep(5) # Wait for 5 seconds
        s.write("serr\n") # Wrong command to stop servo test
    if front[1] < default_position[1]:
        s.write("st\n") # Servo test command
        time.sleep(5) # Wait for 5 seconds
        s.write("serr\n") # Wrong command to stop servo test
    elif front[1] > default_position[1]:
        s.write("st\n") # Servo test command
        time.sleep(5) # Wait for 5 seconds
        s.write("serr\n") # Wrong command to stop servo test
    if left[1] < default_position[2]:
        s.write("st\n") # Servo test command
        time.sleep(5) # Wait for 5 seconds
        s.write("serr\n") # Wrong command to stop servo test
    elif left[1] > default_position[2]:
        s.write("st\n") # Servo test command
        time.sleep(5) # Wait for 5 seconds
        s.write("serr\n") # Wrong command to stop servo test

try:
    sensor_1 = bluez.hci_open_dev(0)
    sensor_2 = bluez.hci_open_dev(1)
    sensor_3 = bluez.hci_open_dev(2)

except:
    print "error accessing bluetooth device..."
    sys.exit(1)

blescan.hci_le_set_scan_parameters(sensor_1)
blescan.hci_enable_le_scan(sensor_1)
blescan.hci_le_set_scan_parameters(sensor_2)
blescan.hci_enable_le_scan(sensor_2)
blescan.hci_le_set_scan_parameters(sensor_3)
blescan.hci_enable_le_scan(sensor_3)

rssi_1 = []
rssi_2 = []
rssi_3 = []

while True:

    output_1 = blescan.parse_events(sensor_1, 1)
    if len(rssi_1) < 20:
        rssi_1.append(output_1)
    else:
        distance_1 = calcDistance(min(rssi_1))
        print distance_1
        rssi_1 = []
    output_2 = blescan.parse_events(sensor_2, 1)
    if len(rssi_2) < 20:
        rssi_2.append(output_2)
    else:
        distance_2 = calcDistance(min(rssi_2))
        print distance_2
        rssi_2 = []
    output_3 = blescan.parse_events(sensor_3, 1)
    if len(rssi_3) < 20:
        rssi_3.append(output_3)
    else:
        distance_3 = calcDistance(min(rssi_3))
        print distance_3
        rssi_3 = []
